### DISCUSSION TITLE
Insert trigger not being respected when inserting via Browse Data grid

### DISCUSSION THREAD

**Discussion by ButtonsTheCalculator** [sqlitebrowser/sqlitebrowser#2626] [Q&A] [answered]
Does anyone know why insert triggers do not fire when inserting records via the Browse Data grid (either by adding a new row or duplicating an existing row)?  I'm trying to ensure that varchar columns' values do not exceed their defined lengths since SQLite does not enforce this.

As a proof of concept, I've added a before update trigger that validates the string length of the value-being-inserted, and I found that updates made both via an update statement in the Execute SQL tab and Write Changes in the Browse Data grid will successfully abort the update with an error message if the length is exceeded (which is the desired behavior).

But creating the same logic for a before insert trigger, I'm only able to get it to fire when inserting via an insert statement in the Execute SQL...doing so via the Write Changes in the Browse Data grid does not fire the trigger.  Unfortunately, the Browse Data grid is our team's usual way of entering data into our database.

Here is the update trigger:
```
CREATE TRIGGER validate_varchar_lengths_update
	BEFORE UPDATE ON myTable
BEGIN
	SELECT
		CASE
			WHEN length(NEW.helpText) > 5 THEN
				RAISE (ABORT, 'Max length exceeded')
		END;
END;
```

...and here is the insert trigger:
```
CREATE TRIGGER validate_varchar_lengths_insert
	BEFORE INSERT ON myTable
BEGIN
	SELECT
		CASE
			WHEN length(NEW.helpText) > 5 THEN
				RAISE (ABORT, 'Max length exceeded')
		END;
END;
```
Any and all help would be greatly appreciated!

**Comment by chrisjlocke** [comment, marked as answer]
Not looked at the trigger logic, but when you insert via the 'Browse Data' grid, there are two actions undertaken - a blank row full of NULLs first, then an edit (for each value entered in a cell) eg:
```
INSERT INTO "main"."filtersets"("filtersetid","brahmsuserid","datatypename","displayname","category","databaseprojectid","addedby","addedon","lastmodifiedby","lastmodifiedon") VALUES (1,NULL,'','',0,NULL,NULL,NULL,NULL,NULL);
UPDATE "main"."filtersets" SET "displayname"=? WHERE "_rowid_"='1'
UPDATE "main"."filtersets" SET "category"=? WHERE "_rowid_"='1'
```
So it would be the 'update' trigger that gets fired on the grid.

**Reply by ButtonsTheCalculator** [reply]
Thank you for the quick response!  Yes, that makes perfect sense and is exactly what I'm seeing now that I've added the update trigger.  Unfortunately, I was testing the two triggers in isolation or else I would've discovered this myself.

If you have any suggestions as to better ways to enforce that the varchar lengths are enforced when inserting/updating records in SQLite, please let me know...otherwise it looks like I have a viable solution, assuming the triggers don't negatively impact performance too much.

**Reply by ButtonsTheCalculator** [reply]
One more thing to mention...I just tried adding a CHECK constraint to the column to see if that would work even better…
`CHECK(length(helpText) <= 5)`
...and that appears to work well, though I don't see a way to provide a custom error message like I can with a trigger raising an ABORT message.  However, this still feels like a cleaner (and more performant?) solution.

**Reply by chrisjlocke** [reply]
Doesn't appear to.  The CHECK is a very basic function that just blocks inserting duff data - by the looks of it, you can only have one CHECK per table/field, but you can have multiple triggers, covering nine areas (before and after insert, update and delete, plus three 'instead of' triggers).

**Reply by chrisjlocke** [reply]
I'd say use triggers, as the custom abort message is beneficial.  Nothing worse than bashing at the keyboard fighting with 'Record failed to insert' just because it contains an erroneous space which you can't see...
